src/app/codegen-data-model-provider/CodegenDataModelProvider_Write.cpp

```diff
@@ -322,7 +319,45 @@ DataModel::ActionReturnStatus CodegenDataModelProvider::WriteAttribute(const Dat
     if (!request.operationFlags.Has(DataModel::OperationFlags::kInternal))
     {
         VerifyOrReturnError(!isReadOnly, Status::UnsupportedWrite);
+    }
 
+    // ACL check for non-internal requests
+    bool checkAcl = !request.operationFlags.Has(DataModel::OperationFlags::kInternal);
+
+    // For chunking, ACL check is not re-done if the previous write was successful for the exact same
+    // path. We apply this everywhere as a shortcut, although realistically this is only for AccessControl cluster
+    if (checkAcl && request.previousSuccessPath.has_value())
+    {
+        checkAcl = ((request.path.mEndpointId != request.previousSuccessPath->mEndpointId) ||
+                    (request.path.mClusterId != request.previousSuccessPath->mClusterId) ||
+                    (request.path.mAttributeId != request.previousSuccessPath->mAttributeId));
+    }
+
+    // If not writable, do unsupported write instead of unsupported-access
+
+    if (checkAcl)
+    {
+        ReturnErrorCodeIf(!request.subjectDescriptor.has_value(), Status::UnsupportedAccess);
+
+        Access::RequestPath requestPath{ .cluster     = request.path.mClusterId,
+                                         .endpoint    = request.path.mEndpointId,
+                                         .requestType = Access::RequestType::kAttributeWriteRequest,
+                                         .entityId    = request.path.mAttributeId };
+        CHIP_ERROR err = Access::GetAccessControl().Check(*request.subjectDescriptor, requestPath,
+                                                          RequiredPrivilege::ForWriteAttribute(request.path));
+
+        if (err != CHIP_NO_ERROR)
+        {
+            ReturnErrorCodeIf((err != CHIP_ERROR_ACCESS_DENIED) && (err != CHIP_ERROR_ACCESS_RESTRICTED_BY_ARL), err);
+
+            // TODO: when wildcard/group writes are supported, handle them to discard rather than fail with status
```

Comment:
Should this todo happen in this PR? Sending a response to a group write seems like something we shouldn't be doing

Reply:
This was copied and pasted from src/app/util/ember-compatibility-functions.cpp which this code emulates.

Reply:
I believe we currently do not handle group writes at all (since this originates from https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/ember-compatibility-functions.cpp#L708 , so the best this PR can do is replicate that.

Reply:
We handle group writes just fine, and have CI for them.

The ember-compatibility-functions comments just need fixing, and these should not duplicate the incorrect claims.

The WriteHandler will drop the response for group writes, so this code is correct in that sense and does not need TODO comments at all.

Wildcard writes, on the other hand, are in fact not supported.

Reply:
Dropped the comment - it seems we are supposed to work with this inside WriteHandlers and it is not a responsibility of the provider to figure out what to do, it just reports the status.